File: apps/mobile/shared/Sources/FlagshipCore/BurnerPairing.swift
import Foundation
import CryptoKit

/// Phone-side of the phone↔desktop-burner pairing — the part that's
/// specific to the burner (the SAS + AEAD reuse `QrRelay`, since the
/// constants are identical).
///
/// The burner shows a QR (`flagship://burner?c=<code>&k=<burnerPkB64url>`)
/// plus an 8-char short code. The phone scans the QR (gets the code AND
/// the pubkey) or types the code (gets only the code, learns the pubkey
/// over the relay via `burner-hello`). Both derive the relay session id
/// from the code identically, so a scan and a typed code converge on the
/// same Durable Object.
///
/// Mirrors apps/burner-mac BurnerPairing (Swift) + the cross-platform
/// vector in apps/com burnerPairingVector.test.ts.
public enum BurnerPairing {

    /// Domain-separation tag for code → session-id (must match the burner).
    public static let sidTag = Data("flagship/burner-sid/v1".utf8)
    public static let codeByteCount = 5

    public struct Scanned: Equatable, Sendable {
        public let codeBytes: Data
        /// Present when scanned from a QR; nil when typed (learned later).
        public let burnerPublicKey: Data?
        public init(codeBytes: Data, burnerPublicKey: Data?) {
            self.codeBytes = codeBytes
            self.burnerPublicKey = burnerPublicKey
        }
    }

    public enum PairError: Error, LocalizedError {
        case badCode
        case badQr(String)
        public var errorDescription: String? {
            switch self {
            case .badCode: return "That code isn't valid — check it on the computer."
            case .badQr(let m): return "QR: \(m)"
            }
        }
    }

    /// The relay session id (`<sid>` in `/burner-pipe/<sid>`).
    public static func sessionId(forCodeBytes code: Data) -> String {
        var h = SHA256()
        h.update(data: sidTag)
        h.update(data: code)
        return String(Base64URL.encode(Data(h.finalize())).prefix(32))
    }

    /// Decode a user-typed short code (case-insensitive, dash/space-tolerant).
    public static func codeBytes(fromHumanCode raw: String) -> Data? {
        let cleaned = raw.uppercased().filter { $0 != " " && $0 != "-" }
        guard let bytes = Base32.decode(cleaned), bytes.count == codeByteCount else { return nil }
        return bytes
    }

    /// Parse either a scanned burner QR or a typed short code.
    /// Accepts: `flagship://burner?c=<code>&k=<pk>`, `c=<code>&k=<pk>`,
    /// or a bare 8-char code.
    public static func parse(_ raw: String) throws -> Scanned {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { throw PairError.badQr("empty") }

        // Bare typed code (no query).
        if !text.contains("=") && !text.contains("?") {
            guard let code = codeBytes(fromHumanCode: text) else { throw PairError.badCode }
            return Scanned(codeBytes: code, burnerPublicKey: nil)
        }

        var c: String?
        var k: String?
        let normalized = text.hasPrefix("flagship://")
            ? text.replacingOccurrences(of: "flagship://", with: "https://_/")
            : text
        if let comps = URLComponents(string: normalized) {
            for q in comps.queryItems ?? [] {
                if q.name == "c" { c = q.value }
                if q.name == "k" { k = q.value }
            }
        }
        guard let codeStr = c, let code = codeBytes(fromHumanCode: codeStr) else {
            throw PairError.badCode
        }
        var pk: Data? = nil
        if let kStr = k, let raw = Base64URL.decode(kStr), raw.count == 32 { pk = raw }
        return Scanned(codeBytes: code, burnerPublicKey: pk)
    }

    /// True if a string looks like a burner pairing QR/code (for the
    /// scanner's `validate` predicate).
    public static func looksLikeBurnerCode(_ raw: String) -> Bool {
        (try? parse(raw)) != nil
    }
}

/// RFC 4648 base32 (uppercase A–Z2–7, no padding). Mirrors apps/burner-mac.
public enum Base32 {
    private static let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")

    public static func encode(_ data: Data) -> String {
        var out = "", buffer = 0, bits = 0
        for byte in data {
            buffer = (buffer << 8) | Int(byte); bits += 8
            while bits >= 5 { bits -= 5; out.append(alphabet[(buffer >> bits) & 0x1f]) }
        }
        if bits > 0 { out.append(alphabet[(buffer << (5 - bits)) & 0x1f]) }
        return out
    }

    public static func decode(_ s: String) -> Data? {
        var lookup = [Character: Int]()
        for (i, c) in alphabet.enumerated() { lookup[c] = i }
        var out = Data(), buffer = 0, bits = 0
        for ch in s {
            guard let v = lookup[ch] else { return nil }
            buffer = (buffer << 5) | v; bits += 5
            if bits >= 8 { bits -= 8; out.append(UInt8((buffer >> bits) & 0xff)) }
        }
        return out
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/BurnerPairing.kt
package com.flagshipserver.app.core

import java.security.MessageDigest

/**
 * Phone-side of the phone↔desktop-burner pairing (Kotlin mirror of the
 * iOS shared BurnerPairing + apps/burner-mac). The SAS + AES-GCM reuse
 * QrRelay/QrSession (identical constants); this only adds the
 * burner-specific short-code → session-id mapping + QR parsing.
 *
 * The burner shows `flagship://burner?c=<code>&k=<burnerPkB64url>` plus an
 * 8-char short code. A scanning phone gets the code AND the pubkey; a
 * typed code gets only the code and learns the pubkey over the relay
 * (`burner-hello`). Both derive the session id from the code identically.
 *
 * Pinned to the cross-platform vector in apps/com burnerPairingVector.test.ts.
 */
object BurnerPairing {
    private val SID_TAG = "flagship/burner-sid/v1".toByteArray()
    const val CODE_BYTE_COUNT = 5

    data class Scanned(val codeBytes: ByteArray, val burnerPublicKey: ByteArray?) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Scanned) return false
            if (!codeBytes.contentEquals(other.codeBytes)) return false
            val a = burnerPublicKey; val b = other.burnerPublicKey
            return if (a == null || b == null) a == null && b == null else a.contentEquals(b)
        }
        override fun hashCode(): Int = codeBytes.contentHashCode()
    }

    /** Relay session id (`<sid>` in `/burner-pipe/<sid>`). */
    fun sessionId(codeBytes: ByteArray): String {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(SID_TAG)
        md.update(codeBytes)
        return Base64URL.encode(md.digest()).take(32)
    }

    /** Decode a typed short code (case-insensitive, dash/space-tolerant). */
    fun codeBytes(humanCode: String): ByteArray? {
        val cleaned = humanCode.uppercase().filter { it != ' ' && it != '-' }
        val bytes = Base32.decode(cleaned) ?: return null
        return if (bytes.size == CODE_BYTE_COUNT) bytes else null
    }

    /**
     * Parse a scanned burner QR or a typed short code. Accepts
     * `flagship://burner?c=<code>&k=<pk>`, `c=<code>&k=<pk>`, or a bare code.
     */
    fun parse(raw: String): Scanned? {
        val text = raw.trim()
        if (text.isEmpty()) return null

        if (!text.contains("=") && !text.contains("?")) {
            val code = codeBytes(text) ?: return null
            return Scanned(code, null)
        }

        val query = text.substringAfter("?", text)
        var c: String? = null
        var k: String? = null
        for (pair in query.split("&")) {
            val kv = pair.split("=", limit = 2)
            if (kv.size != 2) continue
            when (kv[0]) {
                "c" -> c = kv[1]
                "k" -> k = kv[1]
            }
        }
        val codeStr = c ?: return null
        val code = codeBytes(codeStr) ?: return null
        val pk = k?.let { Base64URL.decode(it) }?.takeIf { it.size == 32 }
        return Scanned(code, pk)
    }

    fun looksLikeBurnerCode(raw: String): Boolean = parse(raw) != null
}

/** RFC 4648 base32 (uppercase A–Z2–7, no padding). Mirrors apps/burner-mac. */
object Base32 {
    private const val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

    fun encode(data: ByteArray): String {
        val out = StringBuilder()
        var buffer = 0
        var bits = 0
        for (b in data) {
            buffer = (buffer shl 8) or (b.toInt() and 0xff)
            bits += 8
            while (bits >= 5) { bits -= 5; out.append(ALPHABET[(buffer shr bits) and 0x1f]) }
        }
        if (bits > 0) out.append(ALPHABET[(buffer shl (5 - bits)) and 0x1f])
        return out.toString()
    }

    fun decode(s: String): ByteArray? {
        val out = ArrayList<Byte>()
        var buffer = 0
        var bits = 0
        for (ch in s) {
            val v = ALPHABET.indexOf(ch)
            if (v < 0) return null
            buffer = (buffer shl 5) or v
            bits += 5
            if (bits >= 8) { bits -= 8; out.add(((buffer shr bits) and 0xff).toByte()) }
        }
        return out.toByteArray()
    }
}
